README
NAME
    EAV::XS - Email Address Validator

SYNOPSIS
        use EAV::XS;

        my $eav = EAV::XS->new();

        if ($eav->is_email ('valid@example.com')) {
            print "This is a valid email address.\n";
        } else {
            printf "The email address is not valid: %s\n",
                    $eav->get_error();
        }

DESCRIPTION
    The purpose of this module is a validation of the specified Email
    Address <https://en.wikipedia.org/wiki/Email_address>.

    The core part of the module is written in C and can be found in the
    libeav directory.

    The module conforms to:

    *   RFC 822 <https://tools.ietf.org/html/rfc822> - allows control
        characters.

    *   RFC 5321 <https://tools.ietf.org/html/rfc5321> - does not allow any
        control characters.

    *   RFC 5322 <https://tools.ietf.org/html/rfc5322> - allows some control
        characters and not allows SPACE and TAB characters without
        quoted-pairs.

    *   RFC 6531 <https://tools.ietf.org/html/rfc6531> - allows
        Internationalized Email Addresses encoded in UTF-8. See also RFC
        3629 <https://tools.ietf.org/html/rfc3629>. The RFC 6531 is based on
        the rules of the RFC 5321.

    You may change the behavior of the RFC 6531 mode when building the
    module and enable support of the RFC 20
    <https://tools.ietf.org/html/rfc20> and RFC 5322
    <https://tools.ietf.org/html/rfc5322>. By default, neither RFC 5322 nor
    RFC 20 is enabled.

    The RFC 20 disallows the next characters within local-part: "`", "#",
    "^", "{", "}", "~" and "|". They must be in double-quotes.

    The default behavior of the module also includes the check of:

    *   Special and Reserved domains as mentioned in RFC 6761
        <https://tools.ietf.org/html/rfc6761>

    *   FQDN - if the domain contains only alias and it is not a special or
        reserved domain, then the result is negative, that is, such an email
        address is considered as invalid.

    *   TLD - the module checks that domain is a Top Level Domain (TLD). The
        list of TLDs has been taken from IANA's Root Zone Database
        <https://www.iana.org/domains/root/db>. See the "TLD INFORMATION"
        section below for details.

DEPENDENCIES
    You have to install one of IDN libraries on your choice:

    *   libidn2 <https://github.com/libidn/libidn2>

    *   libidn <https://www.gnu.org/software/libidn/>

    *   libidnkit <https://jprs.co.jp/idn/index-e.html>

    When run Makefile.PL, you will be asked to configure EAV::XS and at this
    stage you may select the IDN library to build with.

    Makefile.PL requirements:

    *   ExtUtils::MakeMaker 5.62 or newer

    *   ExtUtils::PkgConfig 1.16 or newer

  INSTALLING DEPENDENCIES
    On Debian 10+, Ubuntu:

        # libidn2 (recommended)
        sudo apt-get install libidn2-dev

        # libidn (only if you really need it or there is no libidn2)
        sudo apt-get install libidn11-dev

        # If in doubt, try search by keyword libidn
        # or idn on other systems.
        #
        # Then try to install XYZ-dev package, where XYZ
        # is libidn2 or libidn2-0 or similar package name.
        apt-cache search libidn

    To install idnkit, download latest release of idnkit2. Unpack it, then
    build and install:

        tar xjvf idnkit-2.3.tar.bz2
        cd idnkit-2.3
        ./configure
        make
        sudo make install

    By default idnkit2 will be installed in the /usr/local directory. You
    may change it by defining prefix at configure stage, for instance:

        ./configure --prefix=$HOME/local
        make
        make install

    Please, see documentation of idnkit for details.

METHODS
    *   $eav = new ( [%options] )

        Creates a new EAV::XS object, if something goes wrong a message will
        be thrown via croak().

        Possible options includes:

        *   rfc - use this RFC specification. Possible values are: *RFC822*,
            *RFC5321*, *RFC5322* or *RFC6531*. Default is *RFC6531*.

        *   tld_check - enable or disable TLD check. Also, this controls
            FQDN check. Enabled by default.

        *   allow_tld - list of TLD types which considered be good. You have
            to specify this list via logical OR ("|"). Information about
            possible values described below in section "TLD INFORMATION".

            Default value is: *TLD_COUNTRY_CODE* | *TLD_GENERIC* |
            *TLD_GENERIC_RESTRICTED* | *TLD_INFRASTRUCTURE* |
            *TLD_SPONSORED* | *TLD_SPECIAL*.

    *   setup (%options)

        Updates options, see description above.

    *   $yes_no = is_email ( $email )

        Validates the specified email. Returns true if the email is valid,
        otherwise returns false.

    *   $error_message = get_error ()

        Returns an error message for the last email address tested by
        is_email() method.

    *   $lpart = get_lpart ()

        Returns local-part of the email after the is_email method call. If
        the email address is invalid, then get_lpart returns nothing.

    *   $domain = get_domain ()

        Returns domain-part of the email after the is_email method call. If
        the email address is invalid, then get_domain returns nothing. The
        returned value $domain could be an IPv4 address either IPv6 one,
        depending on the specified email address passed to is_email ().

    *   $bool = get_is_ipv4 ()

        Returns whether or not the domain-part of the email contains an IPv4
        address, after the is_email method call. If the email address is
        invalid, then get_is_ipv4 returns false.

    *   $bool = get_is_ipv6 ()

        Returns whether or not the domain-part of the email contains an IPv6
        address, after the is_email method call. If the email address is
        invalid, then get_is_ipv6 returns false.

    *   $bool = get_is_domain ()

        Returns whether or not the domain-part of the email contains an
        domain name, after the is_email method call. If the email address is
        invalid, then get_is_domain returns false.

    *   $tld_type = get_tld_type ()

        Returns TLD type of the email after the is_email method call. If the
        email address is invalid, then get_tld_type returns the
        *EAV::XS::TLD_INVALID* value.

        The TLD type is a value that you pass as the allow_tld option to new
        method. For instance, you may compare the return value of the
        get_tld_type method with the value
        *EAV::XS::TLD_GENERIC_RESTRICTED*.

  TLD INFORMATION
    The current list of all TLDs can be found on IANA Root Zone Database
    <https://www.iana.org/domains/root/db> website.

    The allow_tld option accepts the next values:

    *   TLD_NOT_ASSIGNED - allow not assigned TLDs. On IANA website they are
        listed as "Not assigned" in the "TLD Manager" field.

    *   TLD_COUNTRY_CODE - allow country-code TLDs.

    *   TLD_GENERIC - allow generic TLDs.

    *   TLD_GENERIC_RESTRICTED - allow generic-restricted TLDs.

    *   TLD_INFRASTRUCTURE - allow infrastructure TLDs.

    *   TLD_SPONSORED - allow sponsored TLDs.

    *   TLD_RETIRED - allow retired TLDs. On IANA website they are listed as
        "Retired" in the "TLD Manager" field.

    *   TLD_TEST - allow test TLDs.

    *   TLD_SPECIAL - allow Special & Restricted TLDs. See RFC 2606
        <https://tools.ietf.org/html/rfc2606>, RFC 6761
        <https://tools.ietf.org/html/rfc6761> and RFC 7686
        <https://tools.ietf.org/html/rfc7686> for details. Currently, this
        includes the next TLDs: "test.", "invalid.", "localhost.",
        "example.", "onion." and also Second Level Domains, such as,
        "example.com.", "example.net." and "example.org.".

    For instance, to allow only country-code and generic TLDs you have to
    write this:

        my $eav = EAV::XS->new(
            allow_tld => EAV::XS::TLD_COUNTRY_CODE | EAV::XS::TLD_GENERIC
        );

        if (not $eav->is_email ('test@example.biz')) {
            print ".biz is generic-restricted TLD and not allowed.\n";
        }

SEE ALSO
    References:

    *   RFC 20 <https://tools.ietf.org/html/rfc20>

    *   RFC 822 <https://tools.ietf.org/html/rfc822>

    *   RFC 5321 <https://tools.ietf.org/html/rfc5321>

    *   RFC 5322 <https://tools.ietf.org/html/rfc5322>

    *   RFC 6530 <https://tools.ietf.org/html/rfc6530>

    *   RFC 6531 <https://tools.ietf.org/html/rfc6531>

    Other implementations:

    *   Email::Valid

    *   String::Validator::Email

    *   Data::Validate::Email

    *   Email::Address

    *   Email::IsEmail

AUTHOR
    Vitaliy V. Tokarev, <vitaliy.tokarev@gmail.com>

COPYRIGHT AND LICENSE
    Copyright (c) 2017 Vitaliy V. Tokarev

    All rights reserved.

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions are
    met:

    * Redistributions of source code must retain the above copyright notice,
    this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
    IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
    TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
    PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
    HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
    SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
    TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
    PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
    LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
    NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
    SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

AVAILABILITY
    You can obtain the latest version from
    <https://github.com/gh0stwizard/p5-EAV-XS/>.



Changes
Revision history for Perl extension EAV::XS.

0.7.8  Mon Jun 19 15:31:45 UTC 2023
	- update IANA database

0.7.7  Tue Jun  6 13:39:52 UTC 2023
	- added TLD_ALL flag
	- update IANA database

0.7.6  Sun Mar 19 14:53:23 UTC 2023
	- update IANA database
	- added get_tld_type() method

0.7.5  Thu Mar  9 13:53:46 UTC 2023
	- update IANA database

0.7.4  Sun Dec 18 08:57:35 UTC 2022
	- update IANA database

0.7.3  Wed Dec  7 06:30:49 UTC 2022
	- added notes about installing dependencies, thanks to Gabor Szabo
	- update IANA database

0.7.2  Sat Oct 15 13:50:26 UTC 2022
	- skip return value check of setlocale in tests
	- fixed passing RFC options when generating XS.c

0.7.1  Wed Oct 12 08:35:52 UTC 2022
	- fixed XS.c has generating with incorrect compiler options

0.7.0  Tue Oct 11 13:02:30 UTC 2022
	- update to libeav 1.2.0
	- underscores are prohibited inside of domain part

0.6.9  Sun Oct  9 05:59:37 UTC 2022
	- update IANA database

0.6.8  Wed Jul 19 21:12:46 UTC 2022
	- update IANA database

0.6.7  Fri Dec 31 05:18:27 UTC 2021
	- update IANA database

0.6.6  Sat Dec  4 18:41:57 UTC 2021
	- update IANA database

0.6.5  Mon Nov  1 16:47:25 UTC 2021
	- update IANA database

0.6.4  Wed Oct 27 12:19:28 UTC 2021
	- update IANA database

0.6.3  Thu Oct  7 16:41:39 UTC 2021
	- update IANA database

0.6.2  Tue Sep 28 16:05:18 UTC 2021
	- update IANA database

0.6.1  Sat Jun 26 07:02:17 UTC 2021
	- minor fixes in libeav

0.6.0  Sat Jun 19 11:03:59 UTC 2021
	- update IANA database
	- fixed CR and folding checks of RFC 822

0.5.6  Wed Apr 21 15:28:26 UTC 2021
	- update IANA database

0.5.5  Mon Feb 15 06:52:58 UTC 2021
	- update IANA database

0.5.4  Thu Feb 11 06:29:13 UTC 2021
	- update IANA database

0.5.3  Sat Nov 28 10:33:03 UTC 2020
	- update IANA database
	- added the test "20-dots.t" of dots checking
	- renamed t/11-rfc5322.t -> t/12-rfc5322.t
	- added new method "setup"

0.5.2  Sun Nov 22 02:35:52 UTC 2020
	- update IANA database
	- added repository information to Makefile.PL (thanks to Gabor Szabo)

0.5.1  Sat Oct 17 05:21:03 UTC 2020
	- update IANA database

0.5.0  Wed Jul 29 23:45:02 UTC 2020
	- added support for retired domains

0.4.11 Wed Jul 29 22:02:19 UTC 2020
	- fixed configuration by hands on cygwin
	- update IANA database
	- included README.cygwin

0.4.10 Wed Jul  8 23:44:04 UTC 2020
	- add support for DragonFly
	- update IANA database
	- add support for cygwin

0.4.9  Sat Jun 27 07:36:30 UTC 2020
	- update IANA database

0.4.8  Wed Jun  3 09:51:22 UTC 2020
	- update IANA database

0.4.7  Thu May 28 12:16:07 UTC 2020
	- update IANA database

0.4.6  Tue Apr  7 10:02:52 UTC 2020
	- added null pointer check for get_* functions

0.4.5  Sat Apr  4 04:11:28 UTC 2020
	- fix build on Windows
	- updated IANA database

0.4.4  Thu Mar 26 22:43:25 UTC 2020
	- added support of OpenBSD

0.4.3  Thu Mar 26 07:50:10 UTC 2020
	- fix build on Solaris with gmake
	- fix build with idnkit in libeav

0.4.2  Wed Mar 25 07:18:08 UTC 2020
	- bug fix setlocale utf-8 in tests 16, 17, 18

0.4.1  Tue Mar 24 17:52:28 UTC 2020
	- bug fix RT #132221: Endless loop in Makefile.PL

0.4.0  Tue Mar 24 15:34:33 UTC 2020
	- new methods have been added (get_lpart, get_domain, etc)
	- updated IANA database
	- added new example email-info.pl

0.3.4  Tue Mar 24 01:08:52 UTC 2020
	- add support for NetBSD
	- update libeav upstream

0.3.3  Mon Mar 23 09:27:27 UTC 2020
	- bug fix RT #132187, thanks to Slaven Rezić (SREZIC)
	- libidn workaround added in t/08-file.t
	- update libeav upstream

0.3.2  Sun Mar 22 10:45:33 UTC 2020
	- bug fix RT #132191, thanks to Slaven Rezić (SREZIC)
	- bug fix RT #132192, thanks to Slaven Rezić (SREZIC)

0.3.1  Sat Mar 21 08:17:43 UTC 2020
	- update module description

0.3.0  Sat Mar 21 03:31:15 UTC 2020
	- update libeav upstream
	- update questions in Makefile.PL

0.2.2  Tue Apr 25 10:11:18 UTC 2017
	- rename error() method to get_error()

0.2.1  Mon Apr 24 16:59:57 UTC 2017
	- added support for RFC 7686 ("onion." domain)
	- added support for rest RFCs: 822, 5321, 5322

0.1.2  Sun Apr 23 02:37:54 UTC 2017
	- update tests
	- update libeav with improved error handling

0.1.1  Sat Apr 22 02:47:41 UTC 2017
	- fix issues when building with shared perl
	- update license information
	- update examples

0.1.0  Thu Apr 21 00:00:00 2017
	- initial release


examples/count-domains.pl
#!/usr/bin/perl

#
# Validate email addresses in the specified files and print
# thier statistics per domain basis.
#
# Usage: count-domains.pl FILE1 [ FILE2 FILE3 ... ]
#

use strict;
use warnings;
use open qw(:std :utf8);
use EAV::XS;
use Net::DNS ();


# prototypes
sub parse_file($);
sub check_dns($);


my %dns_cache = ();
my $dns = Net::DNS::Resolver->new();


# stats storage
my %stats;
my $invalid;

my $eav = EAV::XS->new()
    or die "Failed to create object of EAV::XS";

# simplify usage of print()
local $\ = ($^O eq 'MSWin32') ? "\r\n" : "\n";

for my $file (@ARGV) {
    %stats = ();
    $invalid = 0;

    parse_file ($file);

    # print stats
    print "-" x 72;
    print $file;
    print "-" x 72;

    for my $domain (sort keys %stats) {
        print $domain . " " . $stats{ $domain };
    }

    print "INVALID " . $invalid;

}

exit 0;


#
# parse_file: parse the file and count statistics for each email.
#
# Return value: none.
#
sub parse_file($) {
    my ($file) = @_;


    open (my $fh, "<", $file)
        or die "open: $file: $!";

    my $domain = "";

    while (my $email = <$fh>) {
        chomp ($email);

        if ($eav->is_email ($email)) {
            $domain = substr $email, rindex ($email, "@") + 1;

            if (not exists $dns_cache{ $domain }) {
                $dns_cache{ $domain } = check_dns ($domain);
            }

            if ($dns_cache{ $domain }) {
                $stats{ $domain }++;
                next;
            }
        }

        $invalid++;
    }

    close ($fh)
        or die "close: $file: $!";
}


#
# In RFC 5321 has been said that DNS check should follow next steps:
# 1) Use MX records for domain if they exist;
# 2) If there is CNAME records use them and go to step 1;
# 3) If there is no MX records and no CNAME records try to use A/AAAA record;
# 4) If there is no A/AAAA records, then test has failed.
#
# XXX Inside of unstable networks this test might give false-positive results.
# According to RFC 5321 there is must be query repeat attempts 
# if the DNS server does not respond.
#
sub check_dns($) {
    my ($domain) = @_;


    my $reply = $dns->query($domain, "MX");

    if (! $reply) {
        return ($dns->query($domain, "A") || $dns->query($domain, "AAAA"));
    }

    for my $rr (grep { $_->type eq 'MX' } $reply->answer) {
        if ($dns->query($rr->exchange(), "A") ||
            $dns->query($rr->exchange(), "AAAA"))
        {
            return 1;
        }
    }

    for my $rr (grep { $_->type eq 'CNAME' } $reply->answer) {
        # XXX recursion
        if (check_dns ($rr->cname())) {
            return 1;
        }
    }

    return 0;
}


lib/EAV/XS.pm
=pod

=encoding utf-8

=head1 NAME

EAV::XS - Email Address Validator

=head1 SYNOPSIS

    use EAV::XS;

    my $eav = EAV::XS->new();

    if ($eav->is_email ('valid@example.com')) {
        print "This is a valid email address.\n";
    } else {
        printf "The email address is not valid: %s\n",
                $eav->get_error();
    }

=head1 DESCRIPTION

The purpose of this module is a validation of the specified
L<Email Address|https://en.wikipedia.org/wiki/Email_address>.

The core part of the module is written in C and can be
found in the B<libeav> directory.

The module conforms to:

=over 4

=item *

L<RFC 822|https://tools.ietf.org/html/rfc822>
- allows control characters.

=item *

L<RFC 5321|https://tools.ietf.org/html/rfc5321>
- does not allow any control characters.

=item *

L<RFC 5322|https://tools.ietf.org/html/rfc5322>
- allows some control characters and not allows SPACE and TAB
characters without quoted-pairs.

=item *

L<RFC 6531|https://tools.ietf.org/html/rfc6531>
- allows Internationalized Email Addresses encoded in UTF-8.
See also L<RFC 3629|https://tools.ietf.org/html/rfc3629>.
The B<RFC 6531> is based on the rules of the B<RFC 5321>.

=back

You may change the behavior of the B<RFC 6531> mode when
building the module and enable support of the
L<RFC 20|https://tools.ietf.org/html/rfc20> and
L<RFC 5322|https://tools.ietf.org/html/rfc5322>.
By default, neither B<RFC 5322> nor B<RFC 20> is enabled.

The B<RFC 20> disallows the next characters within local-part:
C<`>, C<#>, C<^>, C<{>, C<}>, C<~> and C<|>.
They must be in double-quotes.

The default behavior of the module also includes
the check of:

=over 4

=item *

Special and Reserved domains as mentioned
in L<RFC 6761|https://tools.ietf.org/html/rfc6761>

=item *

FQDN - if the domain contains only alias and it is not
a special or reserved domain, then the result is negative,
that is, such an email address is considered as invalid.

=item *

TLD - the module checks that domain is a Top Level Domain (TLD).
The list of TLDs has been taken from IANA's
L<Root Zone Database|https://www.iana.org/domains/root/db>.
See the L</"TLD INFORMATION"> section below for details.

=back

=cut

package EAV::XS;

use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

our %EXPORT_TAGS = ( 'all' => [ qw(
    RFC822
    RFC5321
    RFC5322
    RFC6531
    TLD_INVALID
    TLD_NOT_ASSIGNED
    TLD_COUNTRY_CODE
    TLD_GENERIC
    TLD_GENERIC_RESTRICTED
    TLD_INFRASTRUCTURE
    TLD_SPONSORED
    TLD_TEST
    TLD_SPECIAL
    TLD_RETIRED
    TLD_ALL
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
our @EXPORT = ();

our $VERSION = "0.7.7";

require XSLoader;
XSLoader::load('EAV::XS', $VERSION);

# RFC specification
use constant RFC822     => 0x00000000;
use constant RFC5321    => 0x00000001;
use constant RFC5322    => 0x00000002;
use constant RFC6531    => 0x00000003;

# allow_tld: this types of TLDs considered as OK
use constant TLD_INVALID            => 0x00000002;
use constant TLD_NOT_ASSIGNED       => 0x00000004;
use constant TLD_COUNTRY_CODE       => 0x00000008;
use constant TLD_GENERIC            => 0x00000010;
use constant TLD_GENERIC_RESTRICTED => 0x00000020;
use constant TLD_INFRASTRUCTURE     => 0x00000040;
use constant TLD_SPONSORED          => 0x00000080;
use constant TLD_TEST               => 0x00000100;
use constant TLD_SPECIAL            => 0x00000200;
use constant TLD_RETIRED            => 0x00000400;
# all flags above OR'ed, except TLD_INVALID
use constant TLD_ALL                => 0x000007fc;

1;
__END__

=head1 DEPENDENCIES

You have to install one of IDN libraries on your choice:

=over 4

=item *

L<libidn2|https://github.com/libidn/libidn2>

=item *

L<libidn|https://www.gnu.org/software/libidn/>

=item *

L<libidnkit|https://jprs.co.jp/idn/index-e.html>

=back

When run Makefile.PL, you will be asked to configure EAV::XS and
at this stage you may select the IDN library to build with.

B<Makefile.PL> requirements:

=over 4

=item *

L<ExtUtils::MakeMaker> B<5.62> or newer

=item *

L<ExtUtils::PkgConfig> B<1.16> or newer

=back

=head2 INSTALLING DEPENDENCIES

On Debian 10+, Ubuntu:

    # libidn2 (recommended)
    sudo apt-get install libidn2-dev

    # libidn (only if you really need it or there is no libidn2)
    sudo apt-get install libidn11-dev

    # If in doubt, try search by keyword libidn
    # or idn on other systems.
    #
    # Then try to install XYZ-dev package, where XYZ
    # is libidn2 or libidn2-0 or similar package name.
    apt-cache search libidn

To install idnkit, download latest release of idnkit2.
Unpack it, then build and install:

    tar xjvf idnkit-2.3.tar.bz2
    cd idnkit-2.3
    ./configure
    make
    sudo make install

By default idnkit2 will be installed in the /usr/local directory.
You may change it by defining prefix at configure stage,
for instance:

    ./configure --prefix=$HOME/local
    make
    make install

Please, see documentation of idnkit for details.

=head1 METHODS

=over 4

=item *

$eav = B<new> ( [%options] )

Creates a new EAV::XS object, if something goes wrong a message will
be thrown via croak().

Possible options includes:

=over 4

=item *

B<rfc> - use this RFC specification. Possible values are: I<RFC822>, 
I<RFC5321>, I<RFC5322> or I<RFC6531>. Default is I<RFC6531>.

=item *

B<tld_check> - enable or disable TLD check. Also, this controls FQDN check.
Enabled by default.

=item *

B<allow_tld> - list of TLD types which considered be good. You have to
specify this list via logical OR ("|"). Information about possible
values described below in section L</"TLD INFORMATION">.

Default value is: I<TLD_COUNTRY_CODE> | I<TLD_GENERIC> | 
I<TLD_GENERIC_RESTRICTED> |
I<TLD_INFRASTRUCTURE> |
I<TLD_SPONSORED> |
I<TLD_SPECIAL>.

=back

=item *

B<setup> (%options)

Updates options, see description above.

=item *

$yes_no = B<is_email> ( $email )

Validates the specified email. Returns true if the email
is valid, otherwise returns false.

=item *

$error_message = B<get_error> ()

Returns an error message for the last email address tested
by B<is_email()> method.

=item *

$lpart = B<get_lpart> ()

Returns local-part of the email I<after> the B<is_email> method call.
If the email address is invalid, then B<get_lpart> returns nothing.

=item *

$domain = B<get_domain> ()

Returns domain-part of the email I<after> the B<is_email> method call.
If the email address is invalid, then B<get_domain> returns nothing.
The returned value $domain could be an IPv4 address either IPv6 one,
depending on the specified email address passed to B<is_email> ().

=item *

$bool = B<get_is_ipv4> ()

Returns whether or not the domain-part of the email contains an IPv4
address, I<after> the B<is_email> method call.
If the email address is invalid, then B<get_is_ipv4> returns false.

=item *

$bool = B<get_is_ipv6> ()

Returns whether or not the domain-part of the email contains an IPv6
address, I<after> the B<is_email> method call.
If the email address is invalid, then B<get_is_ipv6> returns false.

=item *

$bool = B<get_is_domain> ()

Returns whether or not the domain-part of the email contains an domain
name, I<after> the B<is_email> method call.
If the email address is invalid, then B<get_is_domain> returns false.

=item *

$tld_type = B<get_tld_type> ()

Returns TLD type of the email I<after> the B<is_email> method call.
If the email address is invalid or an error has been occured,
then B<get_tld_type> returns the C<EAV::XS::TLD_INVALID> value.

The TLD type is a value that you pass as the C<allow_tld> option
to B<new> method. For instance, you may compare the return value
of the B<get_tld_type> method with the value
C<EAV::XS::TLD_GENERIC_RESTRICTED>.

The list of TLD types may be found in the L</"TLD INFORMATION">
section.

=back

=head2 TLD INFORMATION

The current list of all TLDs can be found on
L<IANA Root Zone Database|https://www.iana.org/domains/root/db> website.

The B<allow_tld> option accepts the next values:

=over 4

=item *

TLD_NOT_ASSIGNED - allow not assigned TLDs. On IANA website they are listed
as "Not assigned" in the "TLD Manager" field.

=item *

TLD_COUNTRY_CODE - allow country-code TLDs.

=item *

TLD_GENERIC - allow generic TLDs.

=item *

TLD_GENERIC_RESTRICTED - allow generic-restricted TLDs.

=item *

TLD_INFRASTRUCTURE - allow infrastructure TLDs.

=item *

TLD_SPONSORED - allow sponsored TLDs.

=item *

TLD_RETIRED - allow retired TLDs. On IANA website they are listed
as "Retired" in the "TLD Manager" field.

=item *

TLD_TEST - allow test TLDs.

=item *

TLD_SPECIAL - allow Special & Restricted TLDs.
See L<RFC 2606|https://tools.ietf.org/html/rfc2606>,
L<RFC 6761|https://tools.ietf.org/html/rfc6761> and
L<RFC 7686|https://tools.ietf.org/html/rfc7686> for details.
Currently, this includes the next TLDs: "test.", "invalid.", 
"localhost.", "example.", "onion." and also Second Level Domains,
such as, "example.com.", "example.net." and "example.org.".

=item *

TLD_ALL - all flags above OR'ed:

  TLD_ALL = TLD_NOT_ASSIGNED |
            TLD_COUNTRY_CODE |
            TLD_GENERIC |
            TLD_GENERIC_RESTRICTED |
            TLD_INFRASTRUCTURE |
            TLD_SPONSORED |
            TLD_TEST |
            TLD_SPECIAL |
            TLD_RETIRED

=back


For instance, to allow only country-code and generic TLDs you
have to write this:

    my $eav = EAV::XS->new(
        allow_tld => EAV::XS::TLD_COUNTRY_CODE | EAV::XS::TLD_GENERIC
    );

    if (not $eav->is_email ('test@example.biz')) {
        print ".biz is generic-restricted TLD and not allowed.\n";
    }


=head1 SEE ALSO

References:

=over 4

=item *

L<RFC 20|https://tools.ietf.org/html/rfc20>

=item *

L<RFC 822|https://tools.ietf.org/html/rfc822>

=item *

L<RFC 5321|https://tools.ietf.org/html/rfc5321>

=item *

L<RFC 5322|https://tools.ietf.org/html/rfc5322>

=item *

L<RFC 6530|https://tools.ietf.org/html/rfc6530>

=item *

L<RFC 6531|https://tools.ietf.org/html/rfc6531>

=back

Other implementations:

=over 4

=item *

L<Email::Valid>

=item *

L<String::Validator::Email>

=item *

L<Data::Validate::Email>

=item *

L<Email::Address>

=item *

L<Email::IsEmail>

=back

=head1 AUTHOR

Vitaliy V. Tokarev, E<lt>vitaliy.tokarev@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2017 Vitaliy V. Tokarev

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=head1 AVAILABILITY

You can obtain the latest version from
L<https://github.com/gh0stwizard/p5-EAV-XS/>.

=cut
